src/west/app/project.py

```diff
@@ -1602,6 +1602,204 @@ def do_run(self, args, user_args):
                 failed.append(project)
         self._handle_failed(args, failed)
 
+GREP_EPILOG = '''
+EXAMPLES
+--------
+
+To get "git grep foo" results from all cloned, active projects:
+
+  west grep foo
+
+To do the same:
+
+- with "git grep --untracked": west grep --untracked foo
+- with "ripgrep":              west grep --tool ripgrep foo
+- with "grep --recursive":     west grep --tool grep foo
+
+To switch the default tool:
+
+- to ripgrep:  west config grep.tool ripgrep
+- to grep:     west config grep.tool grep
+
+GREP TOOLS
+----------
+
+This command runs a "grep tool" to search inside projects.
+Supported tools:
+
+- git-grep (default)
+- ripgrep
+- grep
+
+Set the "grep.tool" configuration option to change the default.
+
+Use "--tool" to switch the tool from its default.
+
+TOOL PATH
+---------
+
+Use --tool-path to override the path to the tool. For example:
+
+  west grep --tool ripgrep --tool-path /my/special/ripgrep
+
+Without --tool-path, the "grep.<TOOL>-path" configuration option
+is checked next. For example, to set the default path to ripgrep:
+
+  west config grep.ripgrep-path /my/special/ripgrep
+
+If the option is not set, "west grep" searches for the tool as follows:
+
+- git-grep: search for "git" (and run as "git grep")
+- ripgrep: search for "rg", then "ripgrep"
+- grep: search for "grep"
+
+TOOL ARGUMENTS
+--------------
+
+The "grep.<TOOL>-args" configuration options, if set, contain arguments
+that are always passed to the tool. The defaults for these are:
+
+- git-grep: (none)
+- ripgrep: (none)
+- grep: "--recursive"
+
+Additional command line arguments not recognized by "west grep" are
+passed to the tool after that. For example:
+
+  west grep --this-option-is-passed-to-the-tool
```

Comment:
So what about `west grep --this-option-is-passed-to-the-tool --project=A`? Who gets `--project`? In other words, is everything after the first unknown argument passed to the tool?

I could probably find the answer myself if I were less lazy but I'm not just asking for myself: I think the doc should clarify this.

Reply:
I think the doc is already clear so maybe you can suggest an edit?

The answer is that you have to use `--` to tell the command line parser "everything after this point is not an option, even if it looks like an option. This is per POSIX.

There is an example of how to pass `--project` to the underlying tool later in the help text.

Reply:
> I did my best to make the for-the-impatient help text clear,

As long as the header of the next section fits in the same screen, the impatient can easily skip to the next section :-) I'm all for conciseness but some completeness level matters to.

> West doesn't recognize --foo, but west grep is documented to pass unrecognized options along to the grep tool. It recognizes --project and therefore "consumes" it, not passing it along.

Thanks for the education, I had no idea that POSIX allowed interleaving consumed and passed-through arguments like this. In fact I didn't even know POSIX cared about this problem in general, I assumed each tool was free to do it the way it wants.

While this help should stay short and neither duplicate POSIX nor encourage this sort of interleaving; it should very briefly state that it is was happens. Not to encourage arguments interleaving (what a terrible idea), but to briefly state what happened when things go wrong. When things go wrong is actually the only moment where users read documentation, isn't it? Trial and error FTW!

I think part of what confused me a bit is the word "Additional" which I misinterpreted as "trailing". So just drop it? also, simply change the example so the unknown option is not at the end anymore:

```suggestion
  west grep --unknown-option-passed-to-the-grep-tool --project=A ...
```

And that's it; the documentation is not even longer!

Reply:
Sorry -- it's not that POSIX is concerning itself with 'consumed' or 'passed-through'. POSIX concerns itself with recommendations for command line programs that delineate between positional arguments and options. A summary is available here: https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html (longopts are kind of a separate GNU extension IIRC, but hopefully in a way that makes sense).

> I think part of what confused me a bit is the word "Additional" which I misinterpreted as "trailing". So just drop it? also, simply change the example so the unknown option is not at the end anymore:

OK, will do, thanks!